File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution(object):
    def nextGreaterElement(self, nums1, nums2):
        """
        :type nums1: List[int]
        :type nums2: List[int]
        :rtype: List[int]
        """
        map = { ele : i  for i , ele in enumerate(nums1)}
        stack = []
        result = [-1] * len(nums1)
        for i in range(len(nums2)-1 , -1 , -1) :


            while stack != [] and nums2[i] > stack[-1] :
                stack.pop()
            if nums2[i] in map :
                if stack != [] :
                    result[map[nums2[i]]] = stack[-1] 
            stack.append(nums2[i]) 
        return result 




        

File: 0496-next-greater-element-i/0496-next-greater-element-i.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Stack<Integer> stack = new Stack<>() ;
        HashMap<Integer,Integer> m = new HashMap<>() ;
        int[] result = new int[nums1.length] ;

        for (int i = nums2.length - 1 ; i>= 0 ; i--){
            int ele = -1 ;
            while ( !stack.isEmpty() && nums2[i] > stack.peek()){
                stack.pop();
            }
            if(!stack.isEmpty()){
                ele = stack.peek() ;
            }




            m.put(nums2[i] , ele) ;
            stack.push(nums2[i]);
        }
        for(int i= 0 ; i < nums1.length ; i++) {
            if(m.containsKey(nums1[i])) {
                result[i] = m.get(nums1[i]) ;
            }
            else 
                result[i] = -1 ;

        }
        return result ;







        
    }
}